pkg/database/all_ops.go

```diff
@@ -82,29 +88,39 @@ func (d *db) ExecAll(req *schema.ExecAllRequest) (*schema.TxHeader, error) {
 					return nil, store.ErrIllegalArguments
 				}
 
-				// check key does not exists or it's already a reference
-				entry, err := d.getAt(EncodeKey(x.Ref.Key), 0, 0, index, tx)
-				if err != nil && err != store.ErrKeyNotFound {
-					return nil, err
-				}
-				if entry != nil && entry.ReferencedBy == nil {
-					return nil, ErrFinalKeyCannotBeConvertedIntoReference
+				if req.NoWait && (x.Ref.AtTx != 0 || !x.Ref.BoundRef) {
+					return nil, ErrNoWaitOperationMustBeSelfContained
 				}
 
-				// reference arguments are converted in regular key value items and then atomically inserted
 				_, exists := kmap[sha256.Sum256(x.Ref.ReferencedKey)]
 
-				if !exists || x.Ref.AtTx > 0 {
-					// check referenced key exists and it's not a reference
-					refEntry, err := d.getAt(EncodeKey(x.Ref.ReferencedKey), x.Ref.AtTx, 0, index, tx)
-					if err != nil {
+				if req.NoWait && !exists {
+					return nil, ErrNoWaitOperationMustBeSelfContained
```

Comment:
With some more descriptive info:
```suggestion
					return nil, fmt.Errorf(
					    "%w: can not create a reference to a key that was not set in the same transaction",
					    ErrNoWaitOperationMustBeSelfContained,
					)
```